ewms_pilot/pilot.py

```diff
@@ -438,91 +480,106 @@ async def _consume_and_reply(
+    in_queue.timeout = _REFRESH_INTERVAL
+    msg_waittime_timeout = timeout_wait_for_first_message or timeout_incoming
 
     # GO!
     total_msg_count = 0
     LOGGER.info(
         "Listening for messages from server to process tasks then send results..."
     )
-    # open pub
-    async with out_queue.open_pub() as pub:
+    #
+    # open pub & sub
+    async with out_queue.open_pub() as pub, in_queue.open_sub_manual_acking() as sub:
         LOGGER.info(f"Processing up to {multitasking} tasks concurrently")
-        # open sub
-        async with in_queue.open_sub_manual_acking() as sub:
+        #
+        # "listener loop" -- get messages and do tasks
+        # intermittently halting to process housekeeping things
+        #
+        msg_waittime_current = 0.0
+        while not listener_loop_exit(
+            task_errors, msg_waittime_current, msg_waittime_timeout
+        ):
+            housekeeper.work(in_queue, sub, pub)
+            #
             # get messages/tasks
-            async for in_msg in sub.iter_messages():
-                total_msg_count += 1
-                LOGGER.info(f"Got a task to process (#{total_msg_count}): {in_msg}")
-
-                if total_msg_count == 1:
-                    utils.chirp_status("Tasking")
-
-                task = asyncio.create_task(
-                    process_msg_task(
-                        in_msg.data,
-                        cmd,
-                        task_timeout,
-                        ftype_to_subproc,
-                        ftype_from_subproc,
-                        file_writer,
-                        file_reader,
-                        staging_dir,
-                        keep_debug_dir,
-                    )
-                )
-                pending[task] = in_msg
-
-                # if we've met max concurrent tasks, wait for the next one to finish
-                while len(pending) >= multitasking:
-                    LOGGER.info("Reached max task concurrency limit, waiting...")
-                    pending, task_errors = await _wait_on_tasks_with_ack(
-                        sub,
-                        pub,
-                        pending,
-                        return_when_all_done=False,
-                        previous_task_errors=task_errors,
-                        # TODO: replace when https://github.com/Observation-Management-Service/MQClient/issues/56
-                        rabbitmq_raw_queues=(
-                            [pub.pub] + list(sub._subs.keys())  # type: ignore[arg-type]
-                            if in_queue._broker_client.NAME.lower() == "rabbitmq"
-                            else None
-                        ),
+            if len(pending) >= multitasking:
+                LOGGER.info("At max task concurrency limit")
+            else:
+                LOGGER.info("Listening for incoming message...")
+                try:
+                    in_msg = await anext(sub.iter_messages())  # -> in_queue.timeout
```

Comment:
I'm not sure what the insides of `sub.iter_messages()` looks like, but it's odd that we're creating the iterator on every loop.  Maybe it's fine?

Reply:
it's technically okay, but it is odd-looking. I could add a non-iterator function to mqclient. Otherwise, I could make an iterator instance and re-instantiate that on `StopAsyncIteration`. Thoughts?

Reply:
The typical pattern is to save the iterator to a variable outside the loop and call `anext(iter)` as needed.  But not sure if there are advantages to creating the iterator again.

Reply:
once `StopAsyncIteration` is raised, the iterator is exhausted so each future call will raise `StopAsyncIteration` again. Recreating the instance restarts the internals